File: ios_app/ios_dependencies/Data/DAO/RealmTopicDAO.swift
//
//  RealmTopicDAO.swift
//  ios_dependencies
//
//  Created by Gabriel Coman on 23/04/2020.
//  Copyright © 2020 Gabriel Coman. All rights reserved.
//

import Foundation
import RememberShared
import RealmSwift

class RealmTopicDAO: RealmDAO, TopicDAO {
    
    let realm: Realm?
    
    init(realm: Realm?) {
        self.realm = realm
    }
    
    func insert(dto___ dto: TopicDTO) {
        let model = RealmTopicDTO(id: dto.id, name: dto.name)
        try? realm?.write {
            realm?.add(model)
        }
    }
    
    func getAll() -> [TopicDTO] {
        let result = realm?.objects(RealmTopicDTO.self)
        return result?.map { $0 } ?? []
    }
}

File: android_dependencies/src/main/java/com/app/dependencies/data/dao/RealmTopicDAO.kt
package com.app.dependencies.data.dao

import com.app.dependencies.data.dto.RealmTopicDTO
import com.app.shared.data.dao.TopicDAO
import com.app.shared.data.dto.TopicDTO

class RealmTopicDAO(override val instance: RealmProvider) : RealmDAO, TopicDAO {

    override fun insert(dto: TopicDTO) {
        val realm = instance()
        val roomDTO = RealmTopicDTO(id = dto.id, name = dto.name)
        realm.beginTransaction()
        realm.insertOrUpdate(roomDTO)
        realm.commitTransaction()
    }

    override fun getAll(): List<TopicDTO> {
        val realm = instance()
        val result = realm.where(RealmTopicDTO::class.java).findAll()
        return realm.copyFromRealm(result).toList()
    }

    override fun delete(topicId: Int) {
        val realm = instance()
        val result = realm.where(RealmTopicDTO::class.java).equalTo("id", topicId).findAll()
        realm.beginTransaction()
        result.deleteAllFromRealm()
        realm.commitTransaction()
    }
}